src/store-api/src/storage/requests.rs

```diff
@@ -14,11 +14,66 @@
 
 use std::fmt::{Display, Formatter};
 
+use common_error::ext::BoxedError;
 use common_recordbatch::OrderOption;
 use datafusion_expr::expr::Expr;
+// Re-export vector types from datatypes to avoid duplication
+pub use datatypes::schema::{VectorDistanceMetric, VectorIndexEngineType};
 use strum::Display;
 
-use crate::storage::SequenceNumber;
+use crate::storage::{ColumnId, SequenceNumber};
+
+/// A hint for KNN vector search.
+#[derive(Debug, Clone, PartialEq)]
+pub struct VectorSearchRequest {
+    /// Column ID of the vector column to search.
+    pub column_id: ColumnId,
+    /// The query vector to search for.
+    pub query_vector: Vec<f32>,
+    /// Number of nearest neighbors to return.
+    pub k: usize,
+    /// Distance metric to use (matches the index metric).
+    pub metric: VectorDistanceMetric,
+}
+
+/// Search results from vector index.
+#[derive(Debug, Clone, PartialEq)]
+pub struct VectorSearchMatches {
+    /// Keys (row offsets in the index).
+    pub keys: Vec<u64>,
+    /// Distances from the query vector.
+    pub distances: Vec<f32>,
+}
+
+/// Trait for vector index engines (HNSW implementations).
+///
+/// This trait defines the interface for pluggable vector index engines.
+/// Implementations (e.g., UsearchEngine) are provided by storage engines like mito2.
+pub trait VectorIndexEngine: Send + Sync {
```

Comment:
people often update the vectors because they ran through new models. Is the index append only or will we support some kind of upsert operations here?

Reply:
Yes, the upsert behaves as it currently does by adding data with the same primary key and time index.